File: src/main/java/com/server/ordering/repository/MyCouponRepository.java
package com.server.ordering.repository;

import com.server.ordering.domain.MyCoupon;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.NoResultException;
import java.util.List;

@Repository
@RequiredArgsConstructor
public class MyCouponRepository {

    private final EntityManager em;

    public void save(MyCoupon myCoupon) {
        em.persist(myCoupon);
    }

    public void findOneWithCustomer(String serialNumber, Long customerId) throws NoResultException {
        em.createQuery("select m from MyCoupon m where m.serialNumber =:serialNumber and m.customer.id =:customerId", MyCoupon.class)
                .setParameter("serialNumber", serialNumber)
                .setParameter("customerId", customerId)
                .getSingleResult();
    }

    public List<MyCoupon> findAll(Long customerId) {
        return em.createQuery("select distinct m from MyCoupon m where m.customer.id =:customerId", MyCoupon.class)
                .setParameter("customerId", customerId)
                .getResultList();
    }

    public void remove(Long couponId) {
        MyCoupon myCoupon = em.find(MyCoupon.class, couponId);
        em.remove(myCoupon);
    }
}

File: src/main/kotlin/com/byeoru/ordering_server/repository/MyCouponRepository.kt
package com.byeoru.ordering_server.repository

import com.byeoru.ordering_server.domain.MyCoupon
import org.springframework.stereotype.Repository
import javax.persistence.EntityManager
import javax.persistence.NoResultException

@Repository
class MyCouponRepository(private val em: EntityManager) {

    fun save(myCoupon: MyCoupon) {
        em.persist(myCoupon)
    }

    @Throws(NoResultException::class)
    fun findOneWithCustomer(serialNumber: String?, customerId: Long?) {
        em.createQuery(
            "select m from MyCoupon m where m.serialNumber =:serialNumber and m.customer.id =:customerId",
            MyCoupon::class.java
        )
            .setParameter("serialNumber", serialNumber)
            .setParameter("customerId", customerId)
            .singleResult
    }

    fun findAll(customerId: Long): List<MyCoupon> {
        return em.createQuery(
            "select distinct m from MyCoupon m where m.customer.id =:customerId",
            MyCoupon::class.java
        )
            .setParameter("customerId", customerId)
            .resultList
    }

    fun remove(couponId: Long) {
        val myCoupon = em.find(MyCoupon::class.java, couponId)
        em.remove(myCoupon)
    }
}